src/core/lib/surface/completion_queue.cc

```diff
@@ -253,6 +253,29 @@ typedef struct cq_pluck_data {
   plucker pluckers[GRPC_MAX_COMPLETION_QUEUE_PLUCKERS];
 } cq_pluck_data;
 
+typedef struct cq_callback_data {
+  /** No actual completed events queue, unlike other types */
+
+  /** Number of pending events (+1 if we're not shutdown) */
+  gpr_atm pending_events;
+
+  /** Counter of how many things have ever been queued on this completion queue
+      useful for avoiding locks to check the queue */
+  gpr_atm things_queued_ever;
+
+  /** 0 initially. 1 once we completed shutting */
+  /* TODO: (sreek) This is not needed since (shutdown == 1) if and only if
+   * (pending_events == 0). So consider removing this in future and use
+   * pending_events */
+  gpr_atm shutdown;
```

Comment:
You do not need the `shutdown` field.  The only time you are using this is in `cq_finish_shutdown_callback()` function in the following statement: 
 `GPR_ASSERT(!gpr_atm_no_barrier_load(&cqd->shutdown)); `

You are not reading it anywhere else. This can be safely removed.

(Note: In pluckable completion queue case, `cqd->shutdown` is read in the `cq->pluck()` function to return `GRPC_QUEUE_SHUTDOWN`)